src/Symfony/Component/Routing/Matcher/UrlMatcher.php

```diff
@@ -259,4 +262,28 @@ protected function createRequest($pathinfo)
             'SCRIPT_NAME' => $this->context->getBaseUrl(),
         ));
     }
+
+    /**
+     * Match all variables enclosed in "{}" and substitute value using related
+     * attribute key.
+     *
+     * @param array $defaults
+     * @param array $attributes
+     */
+    private function resolveDefaults(&$defaults, $attributes)
+    {
+        $enclosingPattern = "/{(\w+)}/";
+
+        foreach ($defaults as $key => $value) {
+            $valueProcessed = preg_replace_callback($enclosingPattern, function ($match) use ($attributes) {
+                if (!isset($attributes[$match[1]])) {
```

Comment:
When we merge attributes and defaults (using mergeDefault legacy method), we can't difference which is coming from attribute and which is coming from defaults but we are only interested in default values.

Maybe the best solution for that it's: first all, evaluate default variables using attributes, then, evaluate again for replace enclosing braces, and for last, merge with attributes.